spec/consensus/light-client.md

```diff
@@ -257,28 +258,60 @@ Towards Lite Client Completeness:
 
 
 ```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
-  }
-  if h2.Header.height == h1.Header.height + 1 {
-    // we have adjacent headers that are not matching (failed
-    // the CheckSupport)
-    // we could submit evidence here
-    return false
-  }
-  pivot := (h1.Header.height + h2.Header.height) / 2
-  hp := Commit(pivot)
-    // ask a full node for header of height pivot
-  Store(hp)  
-    // store header hp locally
-  if Bisection(h1,hp,trustlevel) {
-    // only check right branch if hp is trusted
-    // (otherwise a lot of unnecessary computation may be done)
-    return Bisection(hp,h2,trustlevel)
+func verify(h) {
+  if hasExpired(h) return false
+
+  vp_all := totalVotingPower(h.Header.V) // total sum of voting power of validators in h
+
+  if votingpower_in(signers(h.Commit),h.Header.V) > 2/3 * vp_all {
+        return hasExpired(h)
+  } else {
+        return false
   }
-  else {
-    return false
+}
+
+func Bisection(h1,h2,trustlevel) bool{
+  th := h1 // th is trusted header
+  while th.Header.Height <= h2.Header.height do {
+       // try to move trusted header forward with stored headers
+       // we assume here that iteration will be done in order of header heights
+       ih := th
+       for all stored headers h s.t ih.Header.Height < h.Header.height < h2.Header.height do {
+            if CheckSupport(th,h,trustlevel) {
+                th = h
+            } else if h.Header.height == th.Header.height + 1 {
+                return false // fail to verify succesive headers!
+            } else break // for
+        }
+
+        if CheckSupport(th,h2,trustlevel) {
+            return hasExpired(h2)
+        }
+
+        if h2.Header.height == th.Header.height + 1 {
+            // we have adjacent headers that are not matching (failed the CheckSupport)
+            // we could submit evidence here
+            return false
+        }
+
+        // try to move th
+        endHeight = h2.Header.height
+        foundPivot = false
+        while(!foundPivot) {
+            pivot := (th.Header.height + endHeight) / 2
+            hp := Commit(pivot)
+            if !verify(hp) return false
+            Store(hp)
+
+            if CheckSupport(th,hd,trustlevel) {
+                th = hd
+                foundPivot = true
+            } else if pivot.Header.height == th.Header.height + 1 {
+                return false // fail to verify succesive headers!
+            }
+
+            endHeight = pivot
+        }
```

Comment:
```suggestion
        }
        th = pivot
```

Reply:
We have set th to the header inside the while look. Pivot is just height number.